src/cpp/src/vlm_pipeline.cpp

```diff
@@ -345,358 +213,480 @@ ov::Tensor merge_text_and_image_embeddings_llava(
+        size_t bs = encoded_image.get_shape().at(0);
+        std::vector<size_t> patch_len{target_sizes.size()};
+        std::transform(target_sizes.begin(), target_sizes.end(), patch_len.begin(), [](const ImageSize& height_width) {
+            return height_width.height * height_width.width;
+        });
+        adjust_pos_cache(
+            target_sizes,
+            pipe.m_vlm_config.hidden_size,
+            pipe.m_pos_embed_cache
+        );
+        size_t max_patch_len = *std::max_element(patch_len.begin(), patch_len.end());
+        ov::Tensor key_padding_mask(ov::element::boolean, {bs, max_patch_len});
+        bool* mask_data = key_padding_mask.data<bool>();
+        size_t embed_len = pipe.m_pos_embed_cache.get_shape().at(2);
+        ov::Tensor pos_embed(ov::element::f32, {max_patch_len, bs, embed_len});  // BLD => L * B * D
+        float* pos_embed_data = pos_embed.data<float>();
+        float* cache_data = pipe.m_pos_embed_cache.data<float>();
+        size_t _d0 = pipe.m_pos_embed_cache.get_shape().at(0);
+        size_t _d1 = pipe.m_pos_embed_cache.get_shape().at(1);
+        for (size_t i = 0; i < bs; ++i) {
+            size_t target_h = target_sizes.at(i).height;
+            size_t target_w = target_sizes.at(i).width;
+            for (size_t h_idx = 0; h_idx < target_h; ++h_idx) {
+                for (size_t w_idx = 0; w_idx < target_w; ++w_idx) {
+                    std::copy_n(
+                        cache_data + h_idx * _d1 + w_idx,
+                        embed_len,
+                        pos_embed_data + (h_idx * target_w + w_idx) * bs * embed_len + i * embed_len
+                    );
                 }
             }
+            for (size_t flat = target_h * target_w; flat < max_patch_len; ++flat) {
+                std::fill_n(pos_embed_data + flat * bs * embed_len + i * embed_len, embed_len, 0.0f);
+            }
+            std::fill_n(mask_data + i * max_patch_len, patch_len[i], false);
+            std::fill_n(mask_data + i * max_patch_len + patch_len[i], max_patch_len - patch_len[i], true);
         }
+        pipe.m_resampler.set_tensor("x", encoded_image);  // [N, H*W, old_hidden_size]
+        pipe.m_resampler.set_tensor("pos_embed", pos_embed);  // [H*W, N, new_hidden_size]
+        pipe.m_resampler.set_tensor("key_padding_mask", key_padding_mask);  // [N, H*W]
+        pipe.m_resampler.infer();
+        return pipe.m_resampler.get_output_tensor();  // [N, query_num, new_hidden_size]
     }
+};
+
+VLMPipeline::VLMPipeline(
+    const std::filesystem::path& model_dir,
+    const std::string& device,
+    const ov::AnyMap device_config
+) : m_pimpl{std::make_unique<VLMPipelineImpl>(model_dir, device, device_config)} {}
+
+ov::genai::VLMPipeline::~VLMPipeline() = default;
+
+DecodedResults VLMPipeline::generate(
+    const std::string& prompt,
+    const std::vector<ov::Tensor>& rgbs,
```

Comment:
That's the implementation of the second option. Properties are cast  to this general signature while the sample encourages to use Proprieties. There's just no reason to hide this general function. Someone pedantic may want to avoid minor cast overhead and use that one directly.